app/models/components/course/materials_ability_component.rb

```diff
@@ -0,0 +1,83 @@
+module Course::MaterialsAbilityComponent
+  include AbilityHost::Component
+  extend ActiveSupport::Concern
+
+  def define_permissions
+    if user
+      allow_students_show_materials
+      allow_students_upload_materials
+      allow_staff_manage_materials
+    end
+
+    super
+  end
+
+  private
+
+  def material_all_course_users_hash
+    { folder: course_all_course_users_hash }
+  end
+
+  def material_course_staff_hash
+    { folder: course_staff_hash }
+  end
+
+  def allow_students_show_materials
+    valid_materials.each do |properties|
+      can :read, Course::Material, material_all_course_users_hash.reverse_merge(properties)
+    end
+
+    valid_folders.each do |properties|
+      can :read, Course::Material::Folder, course_all_course_users_hash.reverse_merge(properties)
+    end
+  end
+
+  def allow_students_upload_materials
+    alias_action :new_materials, :upload_materials, to: :upload
+    can :upload, Course::Material::Folder, course_all_course_users_hash.
+      merge(can_student_upload: true)
+    can :manage, Course::Material, creator: user
+  end
+
+  def allow_staff_manage_materials
+    can :manage, Course::Material, material_course_staff_hash
+    can :manage, Course::Material::Folder, course_staff_hash.reverse_merge(manageable_folders)
+    # Root folders are not editable
+    cannot [:edit, :destroy], Course::Material::Folder, parent: nil
+  end
+
+  def valid_folders
+    [
+      {
+        start_at: (Time.min..Time.zone.now),
+        end_at: nil
+      },
+      {
+        start_at: (Time.min..Time.zone.now),
+        end_at: (Time.zone.now..Time.max)
+      }
+    ]
+  end
+
+  def valid_materials
+    [
+      {
+        folder: {
```

Comment:
Make sure it’s properly documented, explain that it’s because more than one condition is possible.